File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysWithKDistinct = function(nums, k) {
    function atmax(nums, k){

    let left = 0;
    let count = 0;
    let unique = 0;
    let hashmap = new Map();

    if(k===0)
    return 0;

    for(let right =0; right<nums.length; right++){
        if(hashmap.has(nums[right]))
        hashmap.set(nums[right], hashmap.get(nums[right])+1);

        else
        {
            hashmap.set(nums[right], 1);
        }

        while(hashmap.size > k){
            hashmap.set(nums[left], hashmap.get(nums[left])-1);
            if(hashmap.get(nums[left])===0)
            hashmap.delete(nums[left]);
            left++;
        }

        count += right - left + 1;
    }
    return count;
    }
    return atmax(nums,k)-atmax(nums,k-1);
};

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.java
class Solution {
    public int calc_sub(int[] nums, int k){
        int high=0;
        int low=0;
        int count=0;
        if(k<=0)
        return 0;
        Map<Integer, Integer> hashmap = new HashMap<>();
        while(high<nums.length){
            hashmap.put(nums[high], hashmap.getOrDefault(nums[high], 0)+1);
            while(hashmap.size()>k){
                hashmap.put(nums[low], hashmap.get(nums[low])-1);
                if(hashmap.get(nums[low])==0)
                hashmap.remove(nums[low]);
                low++;
            }
            count+=(high-low+1);
            high++;
        }
        return count;
    }

    public int subarraysWithKDistinct(int[] nums, int k) {
        return calc_sub(nums,k) - calc_sub(nums, k-1);
        
    }
}